javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -44,8 +44,6 @@
 
     // read http://javarevisited.blogspot.de/2014/05/double-checked-locking-on-singleton-in-java.html
     private transient volatile Supplier<? extends T> supplier;
-
-    // does not need to be volatile, visibility piggy-backs on volatile read of `supplier`
```

Comment:
Maybe this comment is crap but some lines of Lazy (including this) definitely need to be commented.

I'm also no friend of blowing up concise code with _bla, bla_. But here is happening something important behind the curtain. We need to mention that non-volatile instance variables are synced to the main memory when volatile instance variables are updated. Our code strongly relies on this property.

> What does volatile do?
> (...)
> Writing to a volatile field has the same memory effect as a monitor release, and reading from a volatile field has the same memory effect as a monitor acquire. In effect, because the new memory model places stricter constraints on reordering of volatile field accesses with other field accesses, volatile or not, anything that was visible to thread A when it writes to volatile field f becomes visible to thread B when it reads f.

\- [JSR 133 FAQ](https://www.cs.umd.edu/~pugh/java/memoryModel/jsr-133-faq.html)